### PR TITLE
[Core] improve PointLocator and PointOutputProcess

### PR DESCRIPTION
**Pull request by philbucher**
This PR contains two improvements:
- The `PointLocator` by @pablobecker was rewritten to use the computation of the local coordinates provided by the `Geometry`. This way it supports a lot more geometries natively and also the implementation is much simpler. C-Tests are added.
- The `PointOutputProcess` was also improved to work with the reimplemented `PointLocator`. Furthermore it also supports restarts now which was a problem for us in the past. 
It serves as basis for other processes (=> `MultiplePointsOutputProcess` which itself is the basis for the `LineOutputProcess`)
Besides it will also be used for testing in MPI in the future

Tests have been added for each of the processes.

BTW the interface has NOT changed!

@msandre you are already using the `PointOutputProcess` in your tests but I cannot run them because sth is not working also on Master

BTW this PR contains changes in important parts of the core, for these I have created separate PRs: #2298 and #2300

### CODE REVIEW

File: kratos/python_scripts/line_output_process.py
Comment (msandre) [inline kratos/python_scripts/line_output_process.py]: Here number_of_sampling_points should be at least 2. Otherwise you get div0 below.

**Review by msandre — CHANGES_REQUESTED** [review]

**Review by msandre — DISMISSED** [review]

**Comment by philbucher** [comment]
@mpentek please do not merge this yet, I am working on some improvements locally

File: kratos/utilities/point_locator.cpp
Comment (adityaghantasala) [inline kratos/utilities/point_locator.cpp]: Should this be += 1 ?
otherwise checks in the CheckResults (global_objects_found >1) will be not working.

File: kratos/utilities/point_locator.cpp
Comment (adityaghantasala) [inline kratos/utilities/point_locator.cpp]: As a note :  This check is also happening inside the function "IsInside"  (in the geometry base class). This need not be done here.

**Comment by adityaghantasala** [comment]
I am curious where are you using this point locater ? This is brute-force way and is expensive isn"t it ? 
Instead of this why not using the binbased point locater ? (if you have tri and tetra elements)

File: kratos/utilities/point_locator.cpp
Comment (philbucher) [reply kratos/utilities/point_locator.cpp]: no here it is correct the way it is

the name is misleading, it is actually the local number of found entities (cannot be larger than 1 bcs I break after one is found)

In `CheckResults` I do compute the global number then (line 116)

**Comment by philbucher** [comment]
@adityaghantasala I used the same concept as in the original class => `PointLocation`, only this one supports all geometries

I agree that it is brute force and expensive, but this is nothing that is being executed often.

Plus I also need it for Hexas and Quads so I cannot use the binbased point locator

File: kratos/utilities/point_locator.cpp
Comment (adityaghantasala) [reply kratos/utilities/point_locator.cpp]: I see. how about changing the name to local_objects_found ? and defining a different variable global_objects_found in line 116 ? Just to be more clear in implementation.

**Comment by adityaghantasala** [comment]
In this case can we add some comments to the class description saying this point ? 
@philbucher  You are also using this in case of MPI search right ?

File: kratos/utilities/point_locator.cpp
Comment (philbucher) [reply kratos/utilities/point_locator.cpp]: will do :+1:

**Comment by philbucher** [comment]
yep sounds good

File: kratos/utilities/point_locator.cpp
Comment (pooyan-dadvand) [reply kratos/utilities/point_locator.cpp]: You seems to use it in the `CheckResult` method....

**Review by pooyan-dadvand — COMMENTED** [review]
Does PointLocator applies a brute force to find the corresponding objects?!

**Comment by philbucher** [comment]
@pooyan-dadvand yes it is brute force
But it was like this in the original implementation ...

And as I told @adityaghantasala using the BinLocator seems not to work bcs it doesnt support all geometries.
I will check again though

**Comment by philbucher** [comment]
I checked the `BinBasedFastPointLocator` and the usage of Triangular/Tetrahedral elements is very hardcoded (even though I think it would have been little effort making it general)
I will see what I can do, guess it makes sense to improve this one rather than going brute-force :)

**Comment by adityaghantasala** [comment]
@philbucher  From my past understanding : the binbased locater needs the functions "HasIntersection" and "CalculateBoundingBox" and one more function (Name I do not recall exactly) functions to be implemented in all geometry types.  When this is done : one can replace some "configure" functions there to make it general.

**Comment by pooyan-dadvand** [comment]
@philbucher you are right that the BinLocator is written for Tetrahedron/Triangle because it was written long before the required method (like `HasIntersection` and `CalucalateBoundingBox`) were implemented in Kratos. After that we added those method to the geometries but the BinLocator was not updated.

**Comment by philbucher** [comment]
@pooyan-dadvand @adityaghantasala thanks for the explanation.
I didn't see that these other functions are required in the background

In this case I would leave the current version unless someone is against it

**Comment by philbucher** [comment]
now we should be ready with the changes and ready to be re-reviewed

File: kratos/python/add_utilities_to_python.cpp
Comment (RiccardoRossi) [inline kratos/python/add_utilities_to_python.cpp]: why did u change the name?

File: kratos/python_scripts/line_output_process.py
Comment (RiccardoRossi) [inline kratos/python_scripts/line_output_process.py]: how is this fitting with the rest of the PR?

File: kratos/utilities/point_locator.cpp
Comment (RiccardoRossi) [reply kratos/utilities/point_locator.cpp]: this should be in debug, but not in release

File: applications/AdjointFluidApplication/tests/test_vms_sensitivity_2d/cylinder_test_adjoint_parameters.json
Comment (RiccardoRossi) [inline applications/AdjointFluidApplication/tests/test_vms_sensitivity_2d/cylinder_test_adjoint_parameters.json]: i have the impression that this belongs in a different PR...

**Review by RiccardoRossi — COMMENTED** [review]
i have the impression that there are two PRs mixed in this one. on one side there is work done on the output, on the other side there is a change in name in the locator.

could u comment on this?

File: kratos/python/add_utilities_to_python.cpp
Comment (philbucher) [reply kratos/python/add_utilities_to_python.cpp]: 1. to make clear that this is a complete (and more general) reimplementation

2. the new name makes more sense to me -> this calss locates points, it is not itself a location

File: kratos/python_scripts/line_output_process.py
Comment (philbucher) [reply kratos/python_scripts/line_output_process.py]: Will move it to a separate PR

File: applications/AdjointFluidApplication/tests/test_vms_sensitivity_2d/cylinder_test_adjoint_parameters.json
Comment (philbucher) [reply applications/AdjointFluidApplication/tests/test_vms_sensitivity_2d/cylinder_test_adjoint_parameters.json]: Will move to a different PR

File: kratos/utilities/point_locator.cpp
Comment (philbucher) [reply kratos/utilities/point_locator.cpp]: I really think that this should be in release and not only in debug

Otherwise I am silently modifying the bahavior, which will be very hard to debug!

Plus the cost of this is small compared to the computation of this clss

File: kratos/python/add_utilities_to_python.cpp
Comment (msandre) [reply kratos/python/add_utilities_to_python.cpp]: I also prefer the new name.

Comment (RiccardoRossi) [reply kratos/python/add_utilities_to_python.cpp]: but still, by changing the name you are breaking previous usages. that s not a good way to go ...  (i agree the name is better)

File: kratos/utilities/point_locator.cpp
Comment (RiccardoRossi) [reply kratos/utilities/point_locator.cpp]: i am still convinced that these checks belong to debug and not release. 

alsp what about @adityaghantasala comment?

**Comment by RiccardoRossi** [comment]
just to clarify that this class was designed for cases in which you needed to locate one single point on a mesh. if you need multiple points, this should NOT be used (it is wuadratic in cost!!)

File: kratos/python/add_utilities_to_python.cpp
Comment (philbucher) [reply kratos/python/add_utilities_to_python.cpp]: It is not used anywhere else in the code except in the files I modify in this PR

If you prefer the old one I can change it back, your call :)

File: kratos/utilities/point_locator.cpp
Comment (philbucher) [reply kratos/utilities/point_locator.cpp]: @adityaghantasala and I checked, we think it is not applicable

Why should it be done in debug only?

**Comment by philbucher** [comment]
@RiccardoRossi thx for the comments and suggestions

we realize that this is expensive, currently there is nothing else available though

**Comment by philbucher** [comment]
I have revered all the changes and striped this PR down to the minimum
now it is ready to be re-reviewed :)

File: kratos/python_scripts/file_handler_utility.py
Comment (msandre) [inline kratos/python_scripts/file_handler_utility.py]: What about `AsciiFileWriter`. Utility feels like a bit of a noise word to me and the class seems much more specialized than its name suggests.

Comment (msandre) [reply kratos/python_scripts/file_handler_utility.py]: Besides that, I feel like this module is not that _general_ and am a bit concerned about releasing it into the wild. It appears to have at least two responsibilities:

1. Control the flush frequency.
2. Manage restarts.

For point 1, I suggest using the buffering parameter in `open` instead of flush frequency (see https://lyceum-allotments.github.io/2017/03/python-and-pipes-part-4-on-the-buffers/). This will simplify a lot.

If you eliminate the flush frequency, then for point 2 I think you really only need the `AddToExistingOutputFile`. This function make assumptions about the file format which are not enforced by the `FileHandlerUtility`. E.g., header lines begin with `#`, each line starts with a time value. This is most certainly the case for point_output_process but not in general. Maybe this function could be put in the same module as point_output_process and we can eliminate file_handler_utility.py?

Comment (philbucher) [reply kratos/python_scripts/file_handler_utility.py]: Thx for the suggestions

Regarding the name I am very open to improvements.

Flushing seems also better with your suggestion, I will change that

Regarding the restarts: I would like to later also use this for other processes that print data. Therefore I put it to a separate class.

Would it be ok for you if the name indicates the format? Sth like `TimeBasedAsciiFileWriter`? Still not the best name but I think it conveys the idea

Comment (msandre) [reply kratos/python_scripts/file_handler_utility.py]: Ok, in that case, it might make sense to make `TimeBasedAsciiFileWriter` a process and initialize the ascii file in the `ExecuteInitialize` so the restart details are implemented there.

File: kratos/python/add_utilities_to_python.cpp
Comment (RiccardoRossi) [reply kratos/python/add_utilities_to_python.cpp]: No, if it is not used anywhere (including python) then you can change it.

However I would propose to take the occasion to call it something like
**BruteForcePointLocator**

**Comment by RiccardoRossi** [comment]
@philbucher but are u using this in the mapper? i really really would advise against...it is just tooooooo slow (N² !!!!!!! ... i know you are aware of it)

I think it would definitely pay off improving the BinBased Locator (i know that this has been done in a local branch by @roigcarlo and @iiaconeta some time ago, so i don't think it should be terribly difficult. @roigcarlo could you comment on this?)

File: kratos/python/add_utilities_to_python.cpp
Comment (RiccardoRossi) [reply kratos/python/add_utilities_to_python.cpp]: would u agree naming it "BruteForcePointLocator" or smthg of the sort?

Comment (philbucher) [reply kratos/python/add_utilities_to_python.cpp]: Sounds good, will change it

**Comment by philbucher** [comment]
@RiccardoRossi ofc I am not using this in the mapper (there I use proper bin-search as instructed by @roigcarlo )

The BruteForcePointLocator is only used for finding some points to print results.

atm I dont have the time to improve the BinBased Locator, too many other things on my plate...

**Comment by RiccardoRossi** [comment]
oh ok,

the BruteForceLocator is ideal for the point output since if you have few points as it is normally the case than the bin construction is more expensive than doing a few brute force searches

**Comment by roigcarlo** [comment]
@RiccardoRossi I think we uploaded the changes along with the mpm code at some point in the past, but I'll take a look nothing was left behind

**Comment by philbucher** [comment]
@RiccardoRossi @msandre done

**Comment by philbucher** [comment]
ping

File: kratos/python_scripts/time_based_ascii_file_writer_utility.py
Comment (msandre) [inline kratos/python_scripts/time_based_ascii_file_writer_utility.py]: According to Python's `open` documentation:

> A negative buffering means to use the system default, which is usually line buffered for tty devices and fully buffered for other files.

So you don't need the if-statement.

File: kratos/python_scripts/time_based_ascii_file_writer_utility.py
Comment (msandre) [inline kratos/python_scripts/time_based_ascii_file_writer_utility.py:55]: Why not just return `None`? One return value is easier to understand. One can check if the file is valid with `if my_file:`.

File: kratos/python_scripts/time_based_ascii_file_writer_utility.py
Comment (msandre) [inline kratos/python_scripts/time_based_ascii_file_writer_utility.py:10]: This is an improvement but I still think the class could be reduced. Now it appears to provide a wrapper to an opened file at the position ready to write output data. After the constructor returns it basically behaves like a native python file object (with limited functionality). Why not just return the naked file to the user rather than a wrapper class? The functions `write`, `flush` and `close` are redundant and could be eliminated.

**Review by msandre — DISMISSED** [review]
See comments.

**Comment by philbucher** [comment]
@msandre done, makes sense!

Can you reapprove pls :)

**Review by msandre — APPROVED** [review]

**Comment by philbucher** [comment]
thx :+1:

**Pull request merged**
